src/hooks/login/useLedgerLogin.ts

```diff
@@ -195,108 +235,111 @@ export function useLedgerLogin({
-      if (!initialized) {
-        setError(failInitializeErrorText);
-        console.warn(failInitializeErrorText);
-        setIsLoading(false);
-        return;
-      }
-      const accounts = await hwWalletP.getAccounts(
+
+      const accounts = await hwProvider.getAccounts(
         startIndex,
         addressesPerPage
       );
-      const ledgerData = await getLedgerConfiguration(hwWalletP);
+
+      const ledgerData = await getLedgerConfiguration(hwProvider);
       setVersion(ledgerData.version);
       setContractDataEnabled(ledgerData.dataEnabled);
       setAccounts(accounts);
       setIsLoading(false);
     } catch (err) {
-      const { errorMessage, defaultErrorMessage } = getLedgerErrorCodes(err);
-      setError(errorMessage ?? defaultErrorMessage);
-      console.error('error', err);
-      setIsLoading(false);
+      onLoginFailed(err);
     }
+
+    return;
   }
 
   async function onStartLogin() {
     if (isLoggedIn) {
       throw new Error(SECOND_LOGIN_ATTEMPT_ERROR);
     }
+
     setError('');
+
     try {
       setIsLoading(true);
+      await initHWProvider();
+
+      if (!(await isHWProviderInitialized())) {
+        return onLoginFailed(failInitializeErrorText);
+      }
+
       if (ledgerAccount != null) {
-        const hwWalletP = new HWProvider();
-        const initialized = await hwWalletP.init();
-        if (!initialized || !selectedAddress) {
-          console.warn(failInitializeErrorText);
-          return;
+        if (!selectedAddress) {
+          return onLoginFailed(failInitializeErrorText);
         }
 
-        const address = await hwWalletP.login({
+        const address = await hwProvider.login({
           addressIndex: selectedAddress.index.valueOf()
         });
-        setAccountProvider(hwWalletP);
 
         if (!address) {
-          setIsLoading(false);
-          console.warn('Login cancelled.');
-          return;
+          return onLoginFailed('Login cancelled.');
         }
 
         dispatch(
           loginAction({ address, loginMethod: LoginMethodsEnum.ledger })
         );
+
         optionalRedirect({
           callbackRoute,
           onLoginRedirect
         });
       } else {
-        if (accounts?.length > 0) {
-          setShowAddressList(true);
-        } else {
+        if (!accounts?.length) {
           await fetchAccounts();
-          setShowAddressList(true);
         }
+
+        setShowAddressList(true);
       }
     } catch (error) {
-      console.error('error ', error);
-      const { defaultErrorMessage } = getLedgerErrorCodes();
-      setError(defaultErrorMessage);
+      onLoginFailed(error);
     } finally {
       setIsLoading(false);
     }
   }
 
+  useEffect(() => {
+    initHWProvider();
+  }, [hwProvider]);
+
   useEffect(() => {
     fetchAccounts();
-  }, [startIndex]);
+  }, [startIndex, showAddressList, hwProvider]);
+
+  useEffect(() => {
+    if (accounts?.length && !showAddressList) {
```

Comment:
extract to constant. 
better not use accounts?.length directly, but with > 0

Reply:
Resolved